[scudo] Add a method to use a hard-coded page size

cferris1000 opened the pull request:
Currently, only Android supports using a hard-code page size. Make this a bit more generic so any platform that wants to can use this.

In addition, add a getPageSizeLogCached() function since this value is used in release.h and can avoid keeping this value around in objects.

Finally, change some of the release.h page size multiplies to shifts using the new page size log value.

cferris1000 committed:
[scudo] Add a method to use a hard-coded page size

Currently, only Android supports using a hard-code page size. Make
this a bit more generic so any platform that wants to can use this.

In addition, add a getPageSizeLogCached() function since this value
is used in release.h and can avoid keeping this value around in
objects.

Finally, change some of the release.h page size multiplies to shifts
using the new page size log value.

cferris1000 committed:
Run clang-format.

cferris1000 commented:
Ping.

ChiaHungDuan commented on compiler-rt/lib/scudo/standalone/common.h:
```diff
@@ -133,18 +133,39 @@ inline void computePercentage(uptr Numerator, uptr Denominator, uptr *Integral,
 }
 
+inline uptr getPageSizeLogCached() {
+  if (LIKELY(PageSizeLogCached)) {
```
nit: single statement doesn't need braces

cferris1000 replied:
Fixed.

ChiaHungDuan commented on compiler-rt/lib/scudo/standalone/common.h:
```diff
@@ -133,18 +133,39 @@ inline void computePercentage(uptr Numerator, uptr Denominator, uptr *Integral,
+  if (LIKELY(PageSizeLogCached)) {
+    return PageSizeLogCached;
+  }
+  getPageSizeSlow();
```
How about having the comment saying PageSizeLogCached will be set in `getPageSizeSlow` or having a DCHECK after?

cferris1000 replied:
I added a comment and a DCHECK.

ChiaHungDuan commented on compiler-rt/lib/scudo/standalone/combined.h:
```diff
@@ -140,6 +140,11 @@ class Allocator {
   typedef typename QuarantineT::CacheT QuarantineCacheT;
 
   void init() {
+#if !defined(SCUDO_PAGE_SIZE)
```
I would suggest removing the `if !defined` because it's something we want to ensure no matter if it's a constant. OTOH, it will catch the mistake `SCUDO_PAGE_SIZE=0`

cferris1000 replied:
Agreed, so removed.

ChiaHungDuan commented on compiler-rt/lib/scudo/standalone/common.h:
```diff
@@ -133,18 +133,39 @@ inline void computePercentage(uptr Numerator, uptr Denominator, uptr *Integral,
+inline constexpr uptr getPageSizeSlow() { return getPageSizeCached(); }
+
+inline constexpr uptr getPageSizeLogCached() {
+  return static_cast<uptr>(__builtin_ctzl(SCUDO_PAGE_SIZE));
```
How about using `getLog2()`?

cferris1000 replied:
Unfortunately, I can't do that and keep the function constexpr. I want to keep it constexpr so I can add a static_assert in the Android config code to make sure the hard-coded value is being used.

I did change the common.cpp to use getLog2 function.

ChiaHungDuan replied:
I see. I was thinking of marking `getLog2` (and `getLeastSignificantSetBitIndex`) as constexpr but that may forbid us adding some DCHECKs there.

ChiaHungDuan commented on compiler-rt/lib/scudo/standalone/common.cpp:
```diff
@@ -12,13 +12,19 @@
 namespace scudo {
 
-uptr PageSizeCached;
+#if !defined(SCUDO_PAGE_SIZE)
```
Not sure if there's the function ordering issue here. Can we group the cases of SCUDO_PAGE_SIZE together like
```
#if defined(SCUDO_PAGE_SIZE)
...
#else
...
#endif
```

cferris1000 replied:
Do you mean pulling in the cases from common.h here? If so, I wanted to leave the getPageSizeSlow in the header to make it inline.

I did add a comment for getPageSize to indicate where it's coming from. I also modified linux.cpp so that getPageSize will not exist if the page size is hard-coded.

ChiaHungDuan replied:
Sorry I didn't notice that it's common.cpp and common.h (I thought this changes were in the same file). Please ignore this.

cferris1000 committed:
Updates for review comments.

Changed to use getLog2 instead of __builtin_ctzl directly in common.cpp.

Remove braces from single statement if.

Call the CHECK in init regardless of the use of SCUDO_PAGE_SIZE.

Added comment and DCHECK in getPageSizeLogCached when SCUDO_PAGE_SIZE not set.

Added comment in common.cpp about where the getPageSize() function comes from.

Only define getPageSize() in linux.cpp if the page size is not hard-coded.
